### PR TITLE
Support pkce part 3

### PR DESCRIPTION
**Pull request by cviecco**

### CODE REVIEW

File: cmd/keymasterd/idp_oidc.go [hunk 1]
```diff
@@ -158,6 +172,21 @@ func (state *RuntimeState) idpOpenIDCClientCanRedirect(client_id string, redirec
 		if parsedURL.Scheme != "https" {
 			return false, nil
 		}
+		if len(parsedURL.RawQuery) > 0 {
+			return false, nil
+		}
+		if strings.Contains(parsedURL.Path, "..") {
+			return false, nil
+		}
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:181]: Please remove blank line.

File: cmd/keymasterd/idp_oidc.go [hunk 2]
```diff
@@ -158,6 +172,21 @@ func (state *RuntimeState) idpOpenIDCClientCanRedirect(client_id string, redirec
...
+		// if no domains, the matchedRE answer is authoritative
+		if len(client.AllowedRedirectDomains) < 1 {
+			return matchedRE, nil
+		}
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:186]: Please remove blank line.

File: cmd/keymasterd/idp_oidc.go [hunk 3]
```diff
@@ -171,6 +200,100 @@ func (state *RuntimeState) idpOpenIDCClientCanRedirect(client_id string, redirec
...
+func (state *RuntimeState) idpOpenIDCIsCorsOriginAllowed(origin string, clientId string) (bool, error) {
+	parsedURL, err := url.Parse(origin)
+	if err != nil {
+		logger.Debugf(1, "user passed unparsable url as string err = %s", err)
+		return false, nil
+	}
+	if parsedURL.Scheme != "https" {
+		return false, nil
+	}
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:212]: Please remove blank line.

File: cmd/keymasterd/idp_oidc.go [hunk 4]
```diff
@@ -171,6 +200,100 @@ func (state *RuntimeState) idpOpenIDCClientCanRedirect(client_id string, redirec
...
+	}
+	return json.Marshal(encodedKeySet)
+}
+
+func sealEncodeData(plaintext, nonce, key []byte) (string, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return "", err
+	}
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:266]: Please remove blank line.

File: cmd/keymasterd/idp_oidc.go [hunk 5]
```diff
@@ -171,6 +200,100 @@ func (state *RuntimeState) idpOpenIDCClientCanRedirect(client_id string, redirec
...
+	var err error
+	decodedCipherText, err := base64.RawURLEncoding.DecodeString(cipherText)
+	if err != nil {
+		return nil, err
+	}
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:286]: Please remove blank line.

File: cmd/keymasterd/idp_oidc.go [hunk 6]
```diff
@@ -234,6 +357,51 @@ func (state *RuntimeState) idpOpenIDCAuthorizationHandler(w http.ResponseWriter,
...
+			return
+		}
+		serializedKeySet, err := state.encryptKeyAndSerialize(key) //json.Marshal(encodedKeySet)
+		if err != nil {
+			logger.Printf("Error encrpyting PKCE encryption key %v", err)
+			state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+			return
+		}
+		protectedCipherTextKeys = string(serializedKeySet)
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:395]: Please remove blank line.

File: cmd/keymasterd/idp_oidc.go [hunk 7]
```diff
@@ -234,6 +357,51 @@ func (state *RuntimeState) idpOpenIDCAuthorizationHandler(w http.ResponseWriter,
...
+		protectedCipherText, err = sealEncodeData([]byte(jsonEncodedData), []byte(jwtId), key)
+		if err != nil {
+			logger.Printf("Error getting random string %v", err)
+			state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+			return
+		}
+
+	}
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:404]: Please remove blank line.

File: cmd/keymasterd/idp_oidc.go [hunk 8]
```diff
@@ -293,16 +465,59 @@ type userInfoToken struct {
...
 		if client.ClientID != client_id {
 			continue
 		}
-		return client_secret == client.ClientSecret
+		return clientSecret == client.ClientSecret
 	}
 	return false
 }
 
+func (state *RuntimeState) idpOpenIDCClientCanDoPKCEAuth(client_id string) (bool, error) {
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:478]: s/client_id/clientId/

File: cmd/keymasterd/idp_oidc.go [hunk 9]
```diff
@@ -293,16 +465,59 @@ type userInfoToken struct {
...
+	for _, client := range state.Config.OpenIDConnectIDP.Client {
+		if client.ClientID != client_id {
+			continue
+		}
+		return client.ClientSecret == "", nil
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:484]: Please remove blank line.

File: cmd/keymasterd/idp_oidc.go [hunk 10]
```diff
@@ -293,16 +465,59 @@ type userInfoToken struct {
...
+	}
+	return false, nil
+}
+
+func (state *RuntimeState) idpOpenIDCValidCodeVerifier(clientId string, codeVerifier string, codeToken keymasterdCodeToken) bool {
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:490]: Please remove blank line.

File: cmd/keymasterd/idp_oidc.go [hunk 11]
```diff
@@ -293,16 +465,59 @@ type userInfoToken struct {
...
+	if err != nil {
+		return false
+	}
+	var protectedData keymasterdIDPCodeProtectedData
+	err = json.Unmarshal([]byte(plainTextJson), &protectedData)
+	if err != nil {
+		return false
+	}
+	state.logger.Debugf(0, "Protected Data Found=%+v", protectedData)
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:506]: Please remove blank line.

File: cmd/keymasterd/idp_oidc.go [hunk 12]
```diff
@@ -357,17 +572,35 @@ func (state *RuntimeState) idpOpenIDCTokenHandler(w http.ResponseWriter, r *http
...
 		pass = r.Form.Get("client_secret")
-		if len(clientID) < 1 || len(pass) < 1 {
+		if len(pass) < 1 && len(codeVerifier) < 1 {
 			logger.Printf("Cannot get auth credentials in auth request")
 			state.writeFailureResponse(w, r, http.StatusUnauthorized, "")
 			return
 		}
+		if len(clientID) < 1 {
+			// This section is kind of unclear. The original rfc6749 spec
+			// Did not had this mandatory, however given the need for password
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc.go:589]: s/had/have/
s/this/this as/

File: cmd/keymasterd/idp_oidc_test.go [hunk 13]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	state, passwdFile, err := setupValidRuntimeStateSigner(t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.Remove(passwdFile.Name()) // clean up
+	state.pendingOauth2 = make(map[string]pendingAuth2Request)
+	state.Config.Base.AllowedAuthBackendsForWebUI = []string{"password"}
+	state.signerPublicKeyToKeymasterKeys()
+	state.HostIdentity = "localhost"
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:299]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 14]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	/*
+		/// we also need to setup the DB:
+		tmpdir, err := ioutil.TempDir("", "keymasterd")
+		if err != nil {
+			t.Fatal(err)
+		}
+		state.Config.Base.DataDirectory = tmpdir
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:307]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 15]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+		defer os.RemoveAll(tmpdir)
+		err = initDB(state)
+		if err != nil {
+			t.Fatal(err)
+		}
+	*/
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:314]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 16]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	valid_client_id := "valid_client_id"
+	//valid_client_secret := "secret_password"
+	valid_redirect_uri := "https://localhost:12345"
+	clientConfig := OpenIDConnectClientConfig{ClientID: valid_client_id, ClientSecret: "", AllowedRedirectURLRE: []string{"localhost"}}
+	state.Config.OpenIDConnectIDP.Client = append(state.Config.OpenIDConnectIDP.Client, clientConfig)
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:320]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 17]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	// now we add a cookie for auth
+	cookieVal, err := state.setNewAuthCookie(nil, "username", AuthTypePassword)
+	if err != nil {
+		t.Fatal(err)
+	}
+	authCookie := http.Cookie{Name: authCookieName, Value: cookieVal}
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:327]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 18]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	//prepare code challenge
+	var CodeVerifier, _ = cv.CreateCodeVerifier()
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:330]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 19]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	// Create code_challenge with S256 method
+	codeChallenge := CodeVerifier.CodeChallengeS256()
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:333]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 20]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	form := url.Values{}
+	form.Add("scope", "openid")
+	form.Add("response_type", "code")
+	form.Add("client_id", valid_client_id)
+	form.Add("redirect_uri", valid_redirect_uri)
+	form.Add("nonce", "123456789")
+	form.Add("state", "this is my state")
+	form.Add("code_challenge_method", "S256")
+	form.Add("code_challenge", codeChallenge)
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:344]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 21]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	postReq, err := http.NewRequest("POST", idpOpenIDCAuthorizationPath, strings.NewReader(form.Encode()))
+	if err != nil {
+		t.Fatal(err)
+	}
+	postReq.Header.Add("Content-Length", strconv.Itoa(len(form.Encode())))
+	postReq.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:351]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 22]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	postReq.AddCookie(&authCookie)
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:353]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 23]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	t.Logf("%+v", rr)
+	locationText := rr.Header().Get("Location")
+	t.Logf("location=%s", locationText)
+	location, err := url.Parse(locationText)
+	if err != nil {
+		t.Fatal(err)
+	}
+	rCode := location.Query().Get("code")
+	t.Logf("rCode=%s", rCode)
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:367]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 24]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	badVerifierTokenReq.Header.Add("Content-Length", strconv.Itoa(len(badVerifierTokenForm.Encode())))
+	badVerifierTokenReq.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	_, err = checkRequestHandlerCode(badVerifierTokenReq, state.idpOpenIDCTokenHandler, http.StatusUnauthorized)
+	if err != nil {
+		t.Fatal(err)
+	}
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:398]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 25]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	//now we do a token request
+	tokenForm := url.Values{}
+	tokenForm.Add("grant_type", "authorization_code")
+	tokenForm.Add("redirect_uri", valid_redirect_uri)
+	tokenForm.Add("code", rCode)
+	tokenForm.Add("client_id", valid_client_id)
+	tokenForm.Add("code_verifier", CodeVerifier.String())
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:406]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 26]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	tokenReq, err := http.NewRequest("POST", idpOpenIDCTokenPath, strings.NewReader(tokenForm.Encode()))
+	if err != nil {
+		t.Fatal(err)
+	}
+	tokenReq.Header.Add("Content-Length", strconv.Itoa(len(tokenForm.Encode())))
+	tokenReq.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:413]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 27]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	resultAccessToken := accessToken{}
+	body := tokenRR.Result().Body
+	err = json.NewDecoder(body).Decode(&resultAccessToken)
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Logf("resultAccessToken='%+v'", resultAccessToken)
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:425]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 28]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	//now the userinfo
+	userinfoForm := url.Values{}
+	userinfoForm.Add("access_token", resultAccessToken.AccessToken)
+
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:429]: Please remove blank line.

File: cmd/keymasterd/idp_oidc_test.go [hunk 29]
```diff
@@ -254,5 +257,186 @@ func TestIdpOpenIDCClientCanRedirectFilters(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	userinfoReq.Header.Add("Content-Length", strconv.Itoa(len(userinfoForm.Encode())))
+	userinfoReq.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	_, err = checkRequestHandlerCode(userinfoReq, state.idpOpenIDCUserinfoHandler, http.StatusOK)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (rgooch) [inline cmd/keymasterd/idp_oidc_test.go:441]: Please remove blank line.

**Review by rgooch — CHANGES_REQUESTED** [review]

File: cmd/keymasterd/idp_oidc.go [hunk 30]
```diff
@@ -556,13 +817,36 @@ type openidConnectUserInfo struct {
...
 		state.writeFailureResponse(w, r, http.StatusBadRequest,
 			"Invalid Method for Userinfo Handler")
 		return
 	}
 	logger.Debugf(2, "userinfo request=%+v", r)
+
+	origin := r.Header.Get("Origin")
+	if r.Method == "OPTIONS" {
+		if origin == "" {
+			state.writeFailureResponse(w, r, http.StatusBadRequest, "Options MUST contain origin")
```
[the remark this answers is not in the document]
Comment (cviecco) [reply cmd/keymasterd/idp_oidc.go:831]: Good catch.

**Pull request merged**